File: src/ClaudeToolbar.Core/Sessions/SessionEvent.cs
using System.Text.Json;

namespace ClaudeToolbar.Core.Sessions;

public enum SessionEventKind { Start, PromptSubmitted, NeedsAttention, Info, Stopped, Failed, Ended }

/// <summary>
/// <paramref name="RunningAgents"/> is only ever non-zero on a Stopped event: the number of background
/// agents the turn left running, taken from the Stop payload's <c>background_tasks</c>.
/// </summary>
public sealed record SessionEvent(SessionEventKind Kind, string SessionId, string? Cwd, string? Message, string? Detail, int RunningAgents = 0);

/// <summary>Turns a Claude Code hook payload into a session event. Anything we do not track returns null.</summary>
public static class HookEventParser
{
    // Not "idle_prompt": Claude Code sends it a minute after every turn ends ("Claude is waiting for your input"),
    // which is the finished cue again with a false "needs you" attached.
    private static readonly HashSet<string> AttentionTypes = new(StringComparer.Ordinal)
    {
        "permission_prompt", "worker_permission_prompt", "agent_needs_input", "elicitation_dialog", "elicitation_url_dialog",
    };

    /// <summary>Background task types that are agents whose results the session is still waiting for. A dev server is not one.</summary>
    private static readonly HashSet<string> AgentTaskTypes = new(StringComparer.Ordinal)
    {
        "subagent", "local_agent", "remote_agent", "in_process_teammate",
    };

    private static readonly HashSet<string> RunningStatuses = new(StringComparer.Ordinal) { "running", "pending" };

    public static SessionEvent? Parse(string json)
    {
        if (string.IsNullOrWhiteSpace(json)) return null;
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return null;
            var name = Str(root, "hook_event_name");
            var id = Str(root, "session_id");
            if (name is null || string.IsNullOrEmpty(id)) return null;
            var cwd = Str(root, "cwd");
            switch (name)
            {
                case "SessionStart":
                    return new SessionEvent(SessionEventKind.Start, id, cwd, null, Str(root, "source"));
                case "UserPromptSubmit":
                    return new SessionEvent(SessionEventKind.PromptSubmitted, id, cwd, null, null);
                case "Notification":
                    var type = Str(root, "notification_type") ?? string.Empty;
                    var kind = AttentionTypes.Contains(type) ? SessionEventKind.NeedsAttention : SessionEventKind.Info;
                    return new SessionEvent(kind, id, cwd, Str(root, "message"), type);
                case "Stop":
                    return new SessionEvent(SessionEventKind.Stopped, id, cwd, null, null, RunningAgents(root));
                case "StopFailure":
                    return new SessionEvent(SessionEventKind.Failed, id, cwd, null, Str(root, "error_type"));
                case "SessionEnd":
                    return new SessionEvent(SessionEventKind.Ended, id, cwd, null, Str(root, "reason"));
                default:
                    return null;
            }
        }
        catch (JsonException)
        {
            return null;
        }
    }

    /// <summary>
    /// Stop fires when the main agent's turn ends, even while agents it launched in the background are still
    /// working; the payload lists them. Counting them here is what lets the tracker hold "finished" back.
    /// </summary>
    private static int RunningAgents(JsonElement root)
    {
        if (!root.TryGetProperty("background_tasks", out var tasks) || tasks.ValueKind != JsonValueKind.Array) return 0;
        var count = 0;
        foreach (var task in tasks.EnumerateArray())
        {
            if (task.ValueKind != JsonValueKind.Object) continue;
            if (Str(task, "type") is { } type && AgentTaskTypes.Contains(type)
                && Str(task, "status") is { } status && RunningStatuses.Contains(status))
                count++;
        }
        return count;
    }

    private static string? Str(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;
}

File: mac/Sources/ClaudeToolbarCore/Sessions/SessionEvent.swift
import Foundation

public enum SessionEventKind: Sendable { case start, promptSubmitted, needsAttention, info, stopped, failed, ended }

public struct SessionEvent: Equatable, Sendable {
    public let kind: SessionEventKind
    public let sessionId: String
    public let cwd: String?
    public let message: String?
    public let detail: String?
    /// Only ever non-zero on a stopped event: the number of background agents the turn left running,
    /// taken from the Stop payload's `background_tasks`.
    public let runningAgents: Int

    public init(kind: SessionEventKind, sessionId: String, cwd: String?, message: String?, detail: String?, runningAgents: Int = 0) {
        self.kind = kind
        self.sessionId = sessionId
        self.cwd = cwd
        self.message = message
        self.detail = detail
        self.runningAgents = runningAgents
    }
}

/// Turns a Claude Code hook payload into a session event. Anything we do not track returns nil.
public enum HookEventParser {
    // Not "idle_prompt": Claude Code sends it a minute after every turn ends ("Claude is waiting for your input"),
    // which is the finished cue again with a false "needs you" attached.
    private static let attentionTypes: Set<String> = ["permission_prompt", "worker_permission_prompt", "agent_needs_input", "elicitation_dialog", "elicitation_url_dialog"]

    /// Background task types that are agents whose results the session is still waiting for. A dev server is not one.
    private static let agentTaskTypes: Set<String> = ["subagent", "local_agent", "remote_agent", "in_process_teammate"]
    private static let runningStatuses: Set<String> = ["running", "pending"]

    public static func parse(_ json: String) -> SessionEvent? {
        let trimmed = json.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty,
              let object = try? JSONSerialization.jsonObject(with: Data(trimmed.utf8), options: [.fragmentsAllowed]),
              let root = object as? [String: Any],
              let name = root["hook_event_name"] as? String,
              let id = root["session_id"] as? String, !id.isEmpty else { return nil }
        let cwd = root["cwd"] as? String
        switch name {
        case "SessionStart":
            return SessionEvent(kind: .start, sessionId: id, cwd: cwd, message: nil, detail: root["source"] as? String)
        case "UserPromptSubmit":
            return SessionEvent(kind: .promptSubmitted, sessionId: id, cwd: cwd, message: nil, detail: nil)
        case "Notification":
            let type = root["notification_type"] as? String ?? ""
            return SessionEvent(kind: attentionTypes.contains(type) ? .needsAttention : .info, sessionId: id, cwd: cwd,
                                message: root["message"] as? String, detail: type)
        case "Stop":
            return SessionEvent(kind: .stopped, sessionId: id, cwd: cwd, message: nil, detail: nil, runningAgents: runningAgents(root))
        case "StopFailure":
            return SessionEvent(kind: .failed, sessionId: id, cwd: cwd, message: nil, detail: root["error_type"] as? String)
        case "SessionEnd":
            return SessionEvent(kind: .ended, sessionId: id, cwd: cwd, message: nil, detail: root["reason"] as? String)
        default:
            return nil
        }
    }

    /// Stop fires when the main agent's turn ends, even while agents it launched in the background are still
    /// working; the payload lists them. Counting them here is what lets the tracker hold "finished" back.
    private static func runningAgents(_ root: [String: Any]) -> Int {
        guard let tasks = root["background_tasks"] as? [Any] else { return 0 }
        return tasks.reduce(0) { count, task in
            guard let task = task as? [String: Any],
                  let type = task["type"] as? String, agentTaskTypes.contains(type),
                  let status = task["status"] as? String, runningStatuses.contains(status) else { return count }
            return count + 1
        }
    }
}
